dbms/src/DataStreams/AddingDefaultsBlockInputStream.cpp

```diff
@@ -0,0 +1,205 @@
+    if (block_missing_values.empty())
+        return res;
+
+    Block evaluate_block{res};
+    /// remove columns for recalculation
+    for (const auto & column : column_defaults)
+        if (evaluate_block.has(column.first))
+            evaluate_block.erase(column.first);
+
+    evaluateMissingDefaults(evaluate_block, header.getNamesAndTypesList(), column_defaults, context, false);
+
+    std::unordered_map<size_t, MutableColumnPtr> mixed_columns;
+
+    for (const ColumnWithTypeAndName & column_def : evaluate_block)
+    {
+        const String & column_name = column_def.name;
+
+        if (column_defaults.count(column_name) == 0)
+            continue;
+
+        size_t block_column_position = res.getPositionByName(column_name);
+        ColumnWithTypeAndName & column_read = res.getByPosition(block_column_position);
+        const auto & defaults_mask = block_missing_values.getDefaultsBitmask(block_column_position);
+
+        checkCalculated(column_read, column_def, defaults_mask.size());
+
+        if (!defaults_mask.empty())
+        {
+            /// TODO: FixedString
+            if (isColumnedAsNumber(column_read.type) || isDecimal(column_read.type))
+            {
+                MutableColumnPtr column_mixed = (*std::move(column_read.column)).mutate();
+                mixNumberColumns(column_read.type->getTypeId(), column_mixed, column_def.column, defaults_mask);
+                column_read.column = std::move(column_mixed);
+            }
+            else
+            {
+                MutableColumnPtr column_mixed = mixColumns(column_read, column_def, defaults_mask);
+                mixed_columns.emplace(block_column_position, std::move(column_mixed));
+            }
+        }
+    }
+
+    if (!mixed_columns.empty())
+    {
+        /// replace columns saving block structure
+        MutableColumns mutation = res.mutateColumns();
+        for (size_t position = 0; position < mutation.size(); ++position)
+        {
+            auto it = mixed_columns.find(position);
+            if (it != mixed_columns.end())
+                mutation[position] = std::move(it->second);
+        }
+        res.setColumns(std::move(mutation));
+    }
+
+    return res;
+}
+
+void AddingDefaultsBlockInputStream::checkCalculated(const ColumnWithTypeAndName & col_read,
+                                                     const ColumnWithTypeAndName & col_defaults,
+                                                     size_t defaults_needed) const
+{
+    size_t column_size = col_read.column->size();
+
+    if (column_size != col_defaults.column->size())
+        throw Exception("Mismach column sizes while adding defaults", ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH);
+
+    if (column_size < defaults_needed)
+        throw Exception("Unexpected defaults count", ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH);
+
+    if (col_read.type->getTypeId() != col_defaults.type->getTypeId())
```

Comment:
This equality check can yield wrong results when e.g. `DateTime('UTC')` and `DateTime('Europe/Moscow')` are compared. `IDataType::equals` should be used instead